core/src/main/scala/cats/data/Chain.scala

```diff
@@ -333,7 +333,7 @@ sealed abstract class Chain[+A] {
       case Singleton(a) => Singleton((a, 0))
       case Append(left, right) =>
         val leftSize = left.length.toInt
-        Append(left.zipWithIndex, right.zipWithIndex.map { case (a, i) => (a, leftSize + i) })
+        Wrap(left.toVector.zipWithIndex ++ right.toVector.zipWithIndex.map { case (a, i) => (a, leftSize + i) })
```

Comment:
why not:
```scala
case a@Append(_, _) =>
  val it = a.iterator
  Wrap(it.zip(Iterator.iterate(0)(_ + 1)).toList)
```
this way, we only materialize one new list and leverage the existing iterator code. The code above materializes to vector (.toVector), one of them is copies again: (`.map(... leftSize + i) })`, then copies them (.zipWithIndex) then concats them (++).

So, there are like 6 materializations of the values. the iterator approach I suggest above does just one.